### PR TITLE
[mlir] Support ROCDL::ReadlaneOp

### PR DESCRIPTION
**Pull request by knwng**
Support ROCDL::ReadlaneOp to solve https://github.com/ROCm/triton-internal/issues/411.

### CODE REVIEW

**Comment by knwng** [comment]
Hi @krzysz00 ! As discussed before, this PR is to add readlane intrinsic to ROCDL. I'm wondering if it's worthy supporting `writelane` and `readfirstlane` at the same time.

File: mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td [hunk 1]
```diff
@@ -197,6 +197,26 @@ def ROCDL_BallotOp :
...
+  Results<(outs LLVM_Type:$res)>,
+  Arguments<(ins LLVM_Type:$src0,
+                 I32:$src1)> {
+  let summary = "Get the value in the specific lane";
+
+  let description = [{
+    Get the value in lane `src1` from input `src0`.
+  }];
+
+  string llvmBuilder = [{
```
Comment (krzysz00) [inline mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:210]: Can we use the `ROCDL_IntrOp` infrastructure? It'll allow round-tripping if someone gets around to that

File: mlir/test/Target/LLVMIR/rocdl.mlir [hunk 2]
```diff
@@ -118,6 +118,25 @@ llvm.func @rocdl.ballot64(%pred : i1) -> i64 {
   llvm.return %0 : i64
 }
 
+llvm.func @rocdl.readlane(%src0 : f32, %src1: f64, %src2: i32, %src3: vector<2 x f32>) -> f32 {
```
Comment (krzysz00) [inline mlir/test/Target/LLVMIR/rocdl.mlir:121]: In addition to this, can we get a test in `mlir/test/Dialect/LLVMIR/rocdl.mlir` just to make sure the op parses/prints correctly?

**Review by krzysz00 — COMMENTED** [review]
Overall, yeah, let's get this pushed up, just a few notes

File: mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td (hunk 1 above)
Comment (knwng) [reply mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td]: Just want to double-check, https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AMDGPU/VOP2Instructions.td#L794-L796 shows the output and 1st input have the same type, while https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/IR/IntrinsicsAMDGPU.td#L2161-L2163 doesn't have that constraint. I'm wondering if we should reflect this in rocdl.readlane by adding trait `AllTypesMatch<["res", "src0"]>`.

File: mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td [hunk 3]
```diff
@@ -197,6 +197,21 @@ def ROCDL_BallotOp :
...
+                 I32:$src1)> {
+  let results = (outs LLVM_Type:$res);
+  let summary = "Get the value in the specific lane.";
+
+  let description = [{
+    Get the value in lane `src1` from input `src0`.
+  }];
+
+  let assemblyFormat = [{
+    $src0 `,` $src1  attr-dict `:` `(` type($src0) `,` type($src1) `)` `->` type($res)
```
Comment (krzysz00) [inline mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:211]: You could optionally just use `type($res)` or `type($src0), type($src1)` here, but, ... it's a low-level intrinsic wrapper, it doesn't really matter.

Same for not having the parentheses.

Comment (krzysz00) [reply mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:211]: Actually, yeah, no, it's fine

**Review by krzysz00 — APPROVED** [review]
Approved, looks good to me!

**Pull request merged**